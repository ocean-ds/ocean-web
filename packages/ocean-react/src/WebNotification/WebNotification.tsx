import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  useState,
  useRef,
} from 'react';
import { X } from '@useblu/ocean-icons-react';
import classNames from 'classnames';
import useWebNotification from './hook/useWebNotification';

export type Position = 'bottom-left' | 'bottom-right';

export type WebNotificationProps = {
  type?: 'info' | 'positive' | 'warning';
  description?: string;
  isOpen: boolean;
  setIsOpen: Dispatch<SetStateAction<boolean>>;
  action?: () => void;
  actionLabel?: string;
  position?: Position;
  className?: string;
  title?: string;
} & React.ComponentPropsWithoutRef<'div'>;

const WebNotification = React.forwardRef<HTMLDivElement, WebNotificationProps>(
  (
    {
      type = 'info',
      description,
      isOpen,
      setIsOpen,
      action,
      actionLabel,
      position = 'bottom-right',
      className,
      title,
    },
    ref
  ) => {
    const { closeWebNotification, dispatchAction } = useWebNotification({
      type,
      isOpen,
      setIsOpen,
      action,
    });

    const [notificationIndex, setNotificationIndex] = useState<number>(0);
    const notificationRef = useRef<HTMLDivElement | null>(null);

    useEffect(() => {
      if (isOpen) {
        const notifications = Array.from(
          document.querySelectorAll('.ods-web-notification')
        ).reverse();
        const index =
          notifications.findIndex((el) => el === notificationRef.current) + 1;
        setNotificationIndex(index);
      }
    }, [isOpen]);

    return (
      <>
        {isOpen && (
          <div
            className={classNames(
              'ods-web-notification',
              `position-${notificationIndex}`,
              `ods-web-notification__${position}-${
                action ? 'action' : 'default'
              }`,
              className
            )}
            ref={(el) => {
              notificationRef.current = el;
              if (ref) {
                if (typeof ref === 'function') {
                  ref(el);
                } else {
                  (
                    ref as React.MutableRefObject<HTMLDivElement | null>
                  ).current = el;
                }
              }
            }}
            data-testid="web-notification-test"
          >
            <div className="ods-web-notification__content">
              <div className="ods-web-notification__container">
                <div className="ods-web-notification__main-content">
                  <div className="ods-web-notification__title">{title}</div>
                  <div className="ods-web-notification__description">
                    {description}
                  </div>
                </div>
                {action && (
                  <div
                    className="ods-web-notification__action"
                    onClick={dispatchAction}
                    role="button"
                    tabIndex={0}
                  >
                    <div
                      className={`ods-web-notification__action-text-${type}`}
                    >
                      {actionLabel ?? 'action'}
                    </div>
                  </div>
                )}
              </div>
              <div
                className="ods-web-notification__wrapper"
                onClick={closeWebNotification}
                role="button"
                tabIndex={0}
              >
                <X size={16} />
              </div>
            </div>
          </div>
        )}
      </>
    );
  }
);

WebNotification.displayName = 'WebNotification';

export default WebNotification;
