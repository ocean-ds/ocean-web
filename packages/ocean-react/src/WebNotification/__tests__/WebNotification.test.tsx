import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import WebNotification from '../WebNotification';

test('renders element properly', () => {
  render(
    <WebNotification
      isOpen
      className="custom-class"
      description="Hello, this is a notification"
      setIsOpen={jest.fn()}
    />
  );

  expect(screen.getByTestId('web-notification-test')).toMatchSnapshot(`
    <div
      class="ods-web-notification position-1 ods-web-notification__bottom-right-default"
      data-testid="web-notification-test"
    >
      <div class="ods-web-notification__content">
        <div class="ods-web-notification__container">
           <div class="ods-web-notification__title">
            Title
          </div>
          <div class="ods-web-notification__description">
            Hello, this is a notification
          </div>
        </div>
        <div
          class="ods-web-notification__wrapper"
        >
          <svg
            fill="none"
            height="16"
            stroke="currentColor"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18L18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </div>
  `);
});

test('renders element with action', () => {
  render(
    <WebNotification
      isOpen
      className="custom-class"
      description="Hello, this is a notification"
      setIsOpen={jest.fn()}
      action={jest.fn()}
      actionLabel="Action"
    />
  );

  const actionLabel = screen.getByText('Action');

  expect(actionLabel).toBeInTheDocument();
  expect(actionLabel).toHaveTextContent('Action');
});

test('calls action on click', () => {
  const action = jest.fn();

  render(
    <WebNotification
      isOpen
      className="custom-class"
      description="Hello, this is a notification"
      setIsOpen={jest.fn()}
      action={action}
      actionLabel="Perform Action"
    />
  );

  const actionButton = screen.getByText('Perform Action');
  fireEvent.click(actionButton);

  expect(action).toBeCalled();
});

test('calls setIsOpen on close button click', () => {
  const setIsOpen = jest.fn();

  render(
    <WebNotification
      isOpen
      className="custom-class"
      description="Hello, this is a notification"
      setIsOpen={setIsOpen}
    />
  );

  const closeButton = screen
    .getByTestId('web-notification-test')
    .querySelector('svg');
  if (closeButton) {
    fireEvent.click(closeButton);
  }

  expect(setIsOpen).toBeCalledWith(false);
});

test('does not render when isOpen is false', () => {
  render(
    <WebNotification
      isOpen={false}
      className="custom-class"
      description="Hello, this is a notification"
      setIsOpen={jest.fn()}
    />
  );

  expect(screen.queryByTestId('web-notification-test')).not.toBeInTheDocument();
});
