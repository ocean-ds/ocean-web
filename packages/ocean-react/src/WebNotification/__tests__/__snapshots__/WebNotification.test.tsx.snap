// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders element properly: 
    <div
      class="ods-web-notification position-1 ods-web-notification__bottom-right-default"
      data-testid="web-notification-test"
    >
      <div class="ods-web-notification__content">
        <div class="ods-web-notification__container">
           <div class="ods-web-notification__title">
            Title
          </div>
          <div class="ods-web-notification__description">
            Hello, this is a notification
          </div>
        </div>
        <div
          class="ods-web-notification__wrapper"
        >
          <svg
            fill="none"
            height="16"
            stroke="currentColor"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18L18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </div>
   1`] = `
<div
  class="ods-web-notification position-1 ods-web-notification__bottom-right-default custom-class"
  data-testid="web-notification-test"
>
  <div
    class="ods-web-notification__content"
  >
    <div
      class="ods-web-notification__container"
    >
      <div
        class="ods-web-notification__main-content"
      >
        <div
          class="ods-web-notification__title"
        />
        <div
          class="ods-web-notification__description"
        >
          Hello, this is a notification
        </div>
      </div>
    </div>
    <div
      class="ods-web-notification__wrapper"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 20 20"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 0 1 1.414 0L10 8.586l4.293-4.293a1 1 0 1 1 1.414 1.414L11.414 10l4.293 4.293a1 1 0 0 1-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 0 1-1.414-1.414L8.586 10 4.293 5.707a1 1 0 0 1 0-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
